lib/multigrid.cpp

```diff
@@ -1591,22 +1602,156 @@ namespace quda
+    double m_map_generate = 2. / lambda_max;
+    double b_map_generate = -1.;
+
+    // we create batches of near-nulls from `B.size() / vectors_per_set` starting
+    // random vectors
+    int num_vec = static_cast<int>(B.size());
+
+    // orthonormalize
+    if (param.mg_global.pre_orthonormalize) {
+      orthonormalize_vectors(B, num_starting_vectors);
+    }
+
+    for (int s = 0; s < num_starting_vectors; s++) {
+
+      // filter
+      {
+        // P0
+        blas::copy(pk, *B[s]);
+
+        if (filter_iterations > 0) {
+          // P1 = m Ap_0 + b p_0
+          std::swap(pkm1, pk); // p_k -> p_{k - 1}
+          dirac_mdm(Apkm1, pkm1, tmp1, tmp2);
+          blas::axpbyz(m_map_filter, Apkm1, b_map_filter, pkm1, pk);
+
+          if (filter_iterations > 1) {
+            // Enter recursion relation
+            for (int k = 2; k <= filter_iterations; k++) {
+              std::swap(pkm2, pkm1); // p_{k - 1} -> p_{k-2}
+              std::swap(pkm1, pk); // p_k -> p_{k-1}
+              dirac_mdm(Apkm1, pkm1, tmp1, tmp2); // compute A p_{k-1}
+              blas::axpbypczw(2. * m_map_filter, Apkm1, 2. * b_map_filter, pkm1, -1., pkm2, pk);
+
+              // heuristic rescale to keep norms in check...
+              if (k % filter_rescale_freq == 0) {
+                double tmp_nrm2 = blas::norm2(pk);
+                logQuda(QUDA_VERBOSE, "Starting vector %d heuristic rescale at %d old norm2 %e\n", s, k, tmp_nrm2);
+                double tmp_inv_nrm = 1. / sqrt(tmp_nrm2);
+                ax(tmp_inv_nrm, pk);
+                ax(tmp_inv_nrm, pkm1);
+                ax(tmp_inv_nrm, pkm2);
+                ax(tmp_inv_nrm, Apkm1);
+              }
+            }
+          }
+        }
+
+        // print the norm (to check for enhancement, normalize)
+        double nrm2 = blas::norm2(pk);
+        logQuda(QUDA_VERBOSE, "Enhanced norm2 for start %d is %e\n", s, nrm2);
+        ax(1.0 / sqrt(nrm2), pk);
+
+        // This is the first filtered vector
+        blas::copy(*B[s], pk);
+      }
+
+      // Now we generate further vectors by another Chebyshev filter from 0 -> lambda_max
+      // P0 is trivially in place
+
+      // P1
+      std::swap(pkm1, pk); // p_k -> p_{k - 1}
+      dirac_mdm(Apkm1, pkm1, tmp1, tmp2);
+      blas::axpbyz(m_map_generate, Apkm1, b_map_generate, pkm1, pk);
+
+      bool first = true;
+
+      for (int i = s + num_starting_vectors; i < num_vec; i += num_starting_vectors) {
+
+        for (int k = first ? 2 : 0; k < iterations_for_next; k++) {
+          std::swap(pkm2, pkm1); // p_{k - 1} -> p_{k-2}
+          std::swap(pkm1, pk); // p_k -> p_{k-1}
+          dirac_mdm(Apkm1, pkm1, tmp1, tmp2); // compute A p_{k-1}
+          blas::axpbypczw(2. * m_map_generate, Apkm1, 2. * b_map_generate, pkm1, -1., pkm2, pk);
+        }
+
+        blas::copy(*B[i], pk);
+
+        if (first) first = false;
+
+      }
+
+    }
+
+    // print norms
+    if (getVerbosity() >= QUDA_VERBOSE) {
```

Comment:
Could make this a `logQuda`

Reply:
The main reason I didn't was because of line 1740; the `blas::norm2` would get evaluated if I used `logQuda` (I believe)

Reply:
It won’t be evaluated: 
https://github.com/lattice/quda/blob/9bae409b742765563b28939e93fa5c43b03da20b/include/util_quda.h#L130

Reply:
https://github.com/lattice/quda/commit/5ac5288cf79843ba7c64c7cd48859130c48fc6e6